### PR TITLE
[WIP] Control transactions from client process

### LINKED ISSUE #585: Do not make queries in Ecto worker

### PR DESCRIPTION
**Pull request by fishcakez**
Resolves #585, #611.

This branch takes a different approach to I suggested in #585 and keeps the transaction state in the client, instead of in the connection processes - on master the state is kept in the worker. Keeping the state in the connection process created significant complexity for the different drivers (postgrex, mariaex etc) and was starting to result in too many lines of code.

This implementation requires heavy use of the process dictionary, which makes logic hard to follow. I will need to add quite a few transaction tests and I would like to check the approach is ok before continuing.

### CODE REVIEW

**Comment by josevalim** [comment]
:+1:

File: lib/ecto/adapters/sql.ex [hunk 1]
```diff
@@ -434,72 +511,160 @@ defmodule Ecto.Adapters.SQL do
...
-      {worker, counter} ->
-        Process.put(key, {worker, counter - 1})
-    end
-
+  defp do_begin(repo, key, %{transactions: [_|_] = trans, depth: n, module: module} = info,
+  opts, timeout) do
+    savepoint = "ecto_#{n}"
+    n = n + 1
+    query = apply(module, :savepoint, [savepoint])
+    info = %{info | transactions: [next: {:savepoint, savepoint}] ++ trans, depth: n}
```
Comment (josevalim) [inline lib/ecto/adapters/sql.ex]: Why do we need to keep the `trans` history? Wouldn't checking the depth be enough?

Comment (fishcakez) [inline lib/ecto/adapters/sql.ex]: It's a stack, not the history, so transactions are removed on rollback/commit but you may have meant that. Checking the depth is not enough, some other state is required as the current depth could be either `before_begin`, `begin_ok` or `begin_error`. I passed the stack around as I found it easier to understand.

**Comment by fishcakez** [comment]
Ouch those 2 failing tests :(. Fixing those is likely going to add more complexity.

**Comment by josevalim** [comment]
I am closing this. I am almost there with #622, writing the docs now. I have commented the log tests for now, so we can merge this work and fix the log tests once we provide the new log entries.

Thanks @fishcakez, this was a great step for us to get things right. The test cases are also fantastic!
